### PR TITLE
Ali add fallback to player

### PR DESCRIPTION
**Pull request by aliInsightTimer**

### CODE REVIEW

**Comment by aliInsightTimer** [comment]
I have tagged this commit as `0.0.34_fallbackurl`.
So please use this as ref in the pubspec.yaml of the project for testing.

**Comment by aliInsightTimer** [comment]
I have now added a new tag `0.0.35_fallbackurl_extended` which includes the latest changes:
- fallback url resuming at the time the track is broken
- player refactor

File: android/src/main/java/danielr2001/audioplayer/AudioPlayerPlugin.java [hunk 1]
```diff
@@ -122,83 +128,46 @@ private void handleMethodCall(final MethodCall call, final MethodChannel.Result
...
+          setupFallbackUrls(call);
           final boolean repeatMode = call.argument("repeatMode");
           final boolean respectAudioFocus = call.argument("respectAudioFocus");
           final boolean isBackground = call.argument("isBackground");
-          
+          final ForegroundPlayParams foregroundPlayParams = isBackground? null : buildForegroundPlayParams(call);
+          try {
+              this.audioObject = buildAudioObject(isBackground, url, foregroundPlayParams);
+          } catch (Exception e) {
+              this.audioObject = null;
```
Comment (lidemin) [inline android/src/main/java/danielr2001/audioplayer/AudioPlayerPlugin.java]: hi @aliInsightTimer , in what kind of cases, it will throw exceptions? The way it catches all exceptions will swollen the useful informations, instead it will just print the stack trace. which will not be enough for us to spot the issues among all the logs. 

my suggestion is, if the error or exception is something we can predict, please try to be more specific and print the log with more useful information. Such as "can not build audio object because of xxxxxx xxxx".

Further more, we might need to throw another specified exception and jump out from the flow here. Otherwise, app might run into such weird flow which will be harder to debug later. 

Thanks.

File: android/src/main/java/danielr2001/audioplayer/AudioPlayerPlugin.java [hunk 2]
```diff
@@ -73,7 +79,7 @@ public void onServiceConnected(ComponentName className, IBinder service) {
       if (playerMode == PlayerMode.PLAYLIST) {
         tempPlayer.playAll((ArrayList<AudioObject>) audioObjects.clone(), tempIndex);
       } else {
-        tempPlayer.play(audioObject);
+        tempPlayer.play(audioObject, fallbackUrlList, maxAttemptsPerUrl);
```
Comment (lidemin) [inline android/src/main/java/danielr2001/audioplayer/AudioPlayerPlugin.java]: Since we are gonna support Playlist soon, can we put these 2 params into each audio object? @aliInsightTimer

File: example/lib/main.dart [hunk 3]
```diff
@@ -6,6 +6,7 @@ import 'package:path_provider/path_provider.dart';
 import 'player_widget.dart';
 import 'package:http/http.dart';
 
+//const kUrl1 = 'https://staticmp3.insighttimer.com/preview/e1s0e0h4r0z7v9a7c0k9b7e5k9s0b2l0u1d3s8t2-v1.mp3';
```
Comment (lidemin) [inline example/lib/main.dart]: do you mind remove this line? @aliInsightTimer

File: android/src/main/java/danielr2001/audioplayer/AudioPlayerPlugin.java (hunk 1 above)
Comment (aliInsightTimer) [reply android/src/main/java/danielr2001/audioplayer/AudioPlayerPlugin.java]: Hi @lidemin , the `buildAudioObject` method only throws exception when the player is supposed to work in foreground (`isBackground = false`) but the `foregroundPlayParams` is `null`:
`    if(foregroundPlayParams == null) throw new Exception("Foreground params must not be null in foreground mode");
`
Which logs a proper exception.
In our code, however, the foreground params can never be null in foreground mode because of this:
`          final ForegroundPlayParams foregroundPlayParams = isBackground? null : buildForegroundPlayParams(call);
          try {
              this.audioObject = buildAudioObject(isBackground, url, foregroundPlayParams);
          } catch (Exception e) {
              this.audioObject = null;
              e.printStackTrace();
          }`
So the reason for throwing exception inside the `buildAudioObject` method is for sanity checks as every method should check its input and if there is anything wrong it should throw an exception to the upstream rather than swallowing it itself. So here we are now swallowing. In the upstream though, in order to handle the exception, we just print the log as we know that in our current code it is impossible for the exception to happen.
In other words, this exception addresses the developers only (as we are writing a library) and not the users, so that just in case the developer is refactoring the code and forgets to set proper value for foregroundPlayParams, the method throws an exceptions and reminds them.

File: android/src/main/java/danielr2001/audioplayer/AudioPlayerPlugin.java (hunk 2 above)
Comment (aliInsightTimer) [reply android/src/main/java/danielr2001/audioplayer/AudioPlayerPlugin.java]: Yes we can, for the `fallbackUrlList` it makes sense as the fallback url list is related to the `audioObject` being played. For the `maxAttemptsPerUrl`, however, it is more like a play setting which is independent of the audio object being played. 
I will refactor the code and add the fallbackList to the audio object. For the `maxAttemptsPerUrl` if you feel it is still needed to be added to the audio object please let me know.

File: android/src/main/java/danielr2001/audioplayer/AudioPlayerPlugin.java (hunk 1 above)
Comment (aliInsightTimer) [reply android/src/main/java/danielr2001/audioplayer/AudioPlayerPlugin.java]: Hi @lidemin , the `buildAudioObject` method only throws exception when the player is supposed to work in foreground (`isBackground = false`) but the `foregroundPlayParams` is `null`:
`    if(foregroundPlayParams == null) throw new Exception("Foreground params must not be null in foreground mode");
`
Which logs a proper exception.
In our code, however, the foreground params can never be null in foreground mode because of this:
`          final ForegroundPlayParams foregroundPlayParams = isBackground? null : buildForegroundPlayParams(call);
          try {
              this.audioObject = buildAudioObject(isBackground, url, foregroundPlayParams);
          } catch (Exception e) {
              this.audioObject = null;
              e.printStackTrace();
          }`
So the reason for throwing exception inside the `buildAudioObject` method is for sanity checks as every method should check its input and if there is anything wrong it should throw an exception to the upstream rather than swallowing it itself. So here we are now swallowing. In the upstream though, in order to handle the exception, we just print the log as we know that in our current code it is impossible for the exception to happen.
In other words, this exception addresses the developers only (as we are writing a library) and not the users, so that just in case the developer is refactoring the code and forgets to set proper value for foregroundPlayParams, the method throws an exceptions and reminds them.

File: example/lib/main.dart (hunk 3 above)
Comment (aliInsightTimer) [reply example/lib/main.dart]: Sure, I will remove all insight-related stuff from it

**Comment by aliInsightTimer** [comment]
The tag `0.0.36_fallbackurl` includes the latest fallback url update
